AmplifyPlugins/API/Sources/AWSAPIPlugin/Core/AppSyncModelDecoder.swift

```diff
@@ -0,0 +1,66 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import Amplify
+
+/// This decoder is registered and used to detect various data payloads to store
+/// inside an `AppSyncModelProvider` when decoding to the `LazyReference` as a "not yet loaded" Reference. If the data payload
+/// can be decoded to the Model, then the model provider is created as a "loaded" reference.
+public struct AppSyncModelDecoder: ModelProviderDecoder {
+    
+    /// Metadata that contains metadata of a model, specifically the identifiers used to hydrate the model.
+    struct Metadata: Codable {
+        let identifiers: [LazyReferenceIdentifier]
+        let apiName: String?
+        
+        func getIdentifiers() -> [(name: String, value: String)] {
+            identifiers.map { identifier in
+                return (name: identifier.name, value: identifier.value)
+            }
+        }
+    }
+    
+    public static func shouldDecode<ModelType: Model>(modelType: ModelType.Type, decoder: Decoder) -> Bool {
+        if (try? Metadata(from: decoder)) != nil {
+            return true
+        }
+        
+        if (try? ModelType(from: decoder)) != nil {
+            return true
+        }
```

Comment:
yea that's right, we also have this inefficieny in the other decoders as well. We could reduce the pattern of `"shouldDecode" -> if true -> "makeModelProvider" that throws` to just calling `makeModelProvider` that doesn't throw, so it if cannot make the model provider, it will move onto the next ModelDecoder

Reply:
Interesting, we should track and try to tackle that soon.

Reply:
just a short update here, i have optimized this in this PR by consolidating the two methods into one! will update this comment with more details

Reply:
The optimization lies how we are not structuring the return type, the protocol's methods is consolidated into just one `decode` 
```
public protocol ModelProviderDecoder {
    static func decode<ModelType: Model>(modelType: ModelType.Type, decoder: Decoder) -> AnyModelProvider<ModelType>?
}
```
the implementation tries to decode and either returns `nil` or the provider itself!